src/Analyser/MutatingScope.php

```diff
@@ -3551,6 +3551,11 @@ public function filterByTruthyValue(Expr $expr): Scope
 
 		$specifiedTypes = $this->typeSpecifier->specifyTypesInCondition($this, $expr, TypeSpecifierContext::createTruthy());
 		$scope = $this->filterBySpecifiedTypes($specifiedTypes);
+
+		$nativeScope = $this->promoteNativeTypes();
+		$nativeSpecifiedTypes = $this->typeSpecifier->specifyTypesInCondition($nativeScope, $expr, TypeSpecifierContext::createTruthy());
+		$scope->nativeExpressionTypes = $nativeScope->filterBySpecifiedTypes($nativeSpecifiedTypes)->expressionTypes;
+
```

Comment:
Currently I don't have a good idea to solve the performance problem, except making native type related options/rules as opt-in.
We simply have to call the methods twice...